controller/pkg/aclprovider/iptablesprovider.go

```diff
@@ -0,0 +1,299 @@
+	Append(table, chain string, rulespec ...string) error
+	// Insert inserts a rule to a chain of table at the required pos
+	Insert(table, chain string, pos int, rulespec ...string) error
+	// Delete deletes a rule of a chain in the given table
+	Delete(table, chain string, rulespec ...string) error
+	// ListChains lists all the chains associated with a table
+	ListChains(table string) ([]string, error)
+	// ClearChain clears a chain in a table
+	ClearChain(table, chain string) error
+	// DeleteChain deletes a chain in the table. There should be no references to this chain
+	DeleteChain(table, chain string) error
+	// NewChain creates a new chain
+	NewChain(table, chain string) error
+	// Commit will commit changes if it is a batch provider.
+	Commit() error
+}
+
+// BatchProvider uses iptables-restore to program ACLs
+type BatchProvider struct {
+	ipt *iptables.IPTables
+
+	//        TABLE      CHAIN    RULES
+	rules       map[string]map[string][]string
+	batchTables map[string]bool
+
+	sync.Mutex
+}
+
+const (
+	restoreCmd = "iptables-restore"
+)
+
+// NewGoIPTablesProvider returns an IptablesProvider interface based on the go-iptables
+// external package.
+func NewGoIPTablesProvider(batchTables []string) (*BatchProvider, error) {
+	ipt, err := iptables.New()
+	if err != nil {
+		return nil, err
+	}
+	batchTablesMap := map[string]bool{}
+	for _, t := range batchTables {
+		batchTablesMap[t] = true
+	}
+
+	return &BatchProvider{
+		ipt:         ipt,
+		rules:       map[string]map[string][]string{},
+		batchTables: batchTablesMap,
+	}, nil
+}
+
+// Append will append the provided rule to the local cache or call
+// directly the iptables command depending on the table.
+func (b *BatchProvider) Append(table, chain string, rulespec ...string) error {
+	b.Lock()
+	defer b.Unlock()
+
+	if _, ok := b.batchTables[table]; !ok {
+		return b.ipt.Append(table, chain, rulespec...)
+	}
+
+	if _, ok := b.rules[table]; !ok {
+		b.rules[table] = map[string][]string{}
+	}
+
+	if _, ok := b.rules[table][chain]; !ok {
+		b.rules[table][chain] = []string{}
+	}
+
+	rule := strings.Join(rulespec, " ")
+
+	b.rules[table][chain] = append(b.rules[table][chain], rule)
+
+	return nil
+}
+
+// Insert will insert the rule in the corresponding position in the local
+// cache or call the corresponding iptables command, depending on the table.
+func (b *BatchProvider) Insert(table, chain string, pos int, rulespec ...string) error {
+
+	if _, ok := b.batchTables[table]; !ok {
+		return b.ipt.Insert(table, chain, pos, rulespec...)
+	}
+
+	b.Lock()
+	defer b.Unlock()
+
+	if _, ok := b.rules[table]; !ok {
+		b.rules[table] = map[string][]string{}
+	}
+
+	if _, ok := b.rules[table][chain]; !ok {
+		b.rules[table][chain] = []string{}
+	}
+
+	rule := strings.Join(rulespec, " ")
+
+	if pos == 1 {
+		b.rules[table][chain] = append([]string{rule}, b.rules[table][chain]...)
+	} else if pos > len(b.rules[table][chain]) {
+		b.rules[table][chain] = append(b.rules[table][chain], rule)
```

Comment:
Clarification: Converting all insert rules to append since iptables-restore buffer only supports append. Is that the case?

Reply:
correct. The buffer has the explicit sequence that you want and sends only appends.